WorkspaceServer.Tests/PipelineStepTests.cs

```diff
@@ -0,0 +1,185 @@
+        }
+
+        [Fact]
+        public async Task It_returns_same_value_to_concurrent_requests()
+        {
+            var seed = 0;
+            var barrier = new Barrier(3);
+            var producer = new PipelineStep<int>(() =>
+            {
+                barrier.SignalAndWait(2.Seconds());
+                return Task.FromResult(++seed);
+            });
+
+            var values = await  Task.WhenAll( Enumerable.Range(0, 3)
+                .AsParallel()
+                .Select((_) => producer.GetLatestAsync()));
+            values.Should().HaveCount(3).And.OnlyContain(i => i == 1);
+        }
+
+        [Fact]
+        public async Task When_invalidate_while_producing_a_value_the_consumer_waiting_will_wait_for_lastest_production_to_be_finish()
+        {
+            var seed = 0;
+            var consumerBarrier = new Barrier(2);
+            var producerBarrier = new Barrier(2);
+            var producer = new PipelineStep<int>(() =>
+            {
+                producerBarrier.SignalAndWait(2.Seconds());
+                return Task.FromResult(Interlocked.Increment(ref seed));
+            });
+
+            var firstConsumer = Task.Run(() =>
+                {
+                    var task = producer.GetLatestAsync();
+                    consumerBarrier.SignalAndWait(2.Seconds());
+                    return task;
+                }
+            );
+
+            var secondConsumer = Task.Run(() =>
+                {
+                    consumerBarrier.SignalAndWait(2.Seconds());
+                    producer.Invalidate();
+                    producerBarrier.RemoveParticipant();
+                    var task = producer.GetLatestAsync();
+                    return task;
+                }
+            );
+
+            var values = await Task.WhenAll(firstConsumer, secondConsumer);
+            values.Should().HaveCount(2).And.OnlyContain(i => i == 2);
+           
+        }
+
+        [Fact]public async Task Sequence_of_steps_produce_a_value()
+        {
+            var step1 = new PipelineStep<int>(()=>Task.FromResult(1));
+            var step2 = step1.Then( (number) => Task.FromResult($"{number} {number}") );
+            var value1 = await step2.GetLatestAsync();
+            value1.Should().Be("1 1");
+        }
+
+        [Fact]
+        public async Task Invalidating_a_step_in_a_sequence_causes_only_that_step_to_re_evaluate()
+        {
+            var seed1 = 0;
+            var step1 = new PipelineStep<int>(() => Task.FromResult(Interlocked.Increment(ref seed1)));
+            var seed2 = 0;
+            var step2 = step1.Then((number) => Task.FromResult($"{number} {Interlocked.Increment(ref seed2)}"));
+            await step2.GetLatestAsync();
+        
+            step2.Invalidate();
+            var value = await step2.GetLatestAsync();
+            value.Should().Be("1 2");
+        }
+
+        [Fact]
+        public async Task Invalidating_a_step_in_a_sequence_causes_all_successor_to_evaluate()
```

Comment:
Given these semantics, shouldn't the value producer take a cancellationToken?

Reply:
Since multiple callers can all be waiting on the same unit of work, who would be allowd to cancel?

Reply:
As we discussed, this is probably a perf optimization for invalidate, where all the successors being invalidated also have any inflight work cancelled.